Capture output of stdout and stderr to Actions

bswinnerton opened the pull request:
This pull request captures the output of both stdout and stderr and makes it accessible to the GitHub Actions metadata for use in other Actions. As described in #29, my use case here is to take the results of an ansible-playbook run and post it as a pull request comment. Most of the logic for this pull request was borrowed from the [Actions Toolkit documentation](https://github.com/actions/toolkit/tree/f0b00fd201c7ddf14e1572a10d5fb4577c4bd6a2/packages/exec#outputoptions).

My understanding of Ansible logging is that it will write the output to stdout ([source](https://docs.ansible.com/ansible/latest/reference_appendices/logging.html#logging-ansible-output)), but I thought it might be a good idea to capture stderr as well. I'm happy to remove that option if that's preferable.

Given my unfamiliarity with this Action, I don't have a great sense of how to test this pull request and could use support from someone who is more familiar 🙏.

##

Resolves https://github.com/dawidd6/action-ansible-playbook/issues/29.

dawidd6 commented:
Just test this here on CI or run locally like this:
```sh
INPUT_PLAYBOOK=test.yml node main.js
```

dawidd6 commented:
Maybe instead of setting the output consider redirecting it to a file. Please bear in mind that we still want to see the logs printed to console.

bswinnerton commented:
Thanks for the tip on CI, @dawidd6. That's very helpful.

I made a series of updates in d82d4c8...e8ab09a. Happy to squash those if you'd like. They make the following changes:

1. Remove the Typescript syntax copied from the docs (whoops 🙈)
2. Removed semicolons to match the rest of the codebase
3. Broke out the arguments to `exec.exec` to separate the command from its parameters and options.

After doing so, I can see the following outputted locally:

```
$ INPUT_PLAYBOOK=test/playbook.yml node main.js
[command]/Users/brooks/.dotfiles/bin/ansible-playbook test/playbook.yml
::set-output name=stderr::[WARNING]: No inventory was parsed, only implicit localhost is available%0A
[WARNING]: No inventory was parsed, only implicit localhost is available
::set-output name=stderr::[WARNING]: provided hosts list is empty, only localhost is available. Note that%0Athe implicit localhost does not match 'all'%0A
[WARNING]: provided hosts list is empty, only localhost is available. Note that
the implicit localhost does not match 'all'
::set-output name=stdout::%0APLAY [Test Action] *************************************************************%0Askipping: no hosts matched%0A

PLAY [Test Action] *************************************************************
skipping: no hosts matched
::set-output name=stdout::%0APLAY RECAP *********************************************************************%0A%0A

PLAY RECAP *********************************************************************

```


```
$ git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
$ INPUT_PLAYBOOK=test/playbook.yml node main.js
[command]/Users/brooks/.dotfiles/bin/ansible-playbook test/playbook.yml
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that
the implicit localhost does not match 'all'

PLAY [Test Action] *************************************************************
skipping: no hosts matched

PLAY RECAP *********************************************************************

```

It's a bet messy locally, but I think that's because of the `::set-output` capture. I'd be curious what this looks like in CI, I bet that Actions parses that and strips it. Unfortunately I don't have the ability to run code in CI since this pull request is coming from a fork targeting the base branch.

dawidd6 commented:
I think you need to append the output in listeners. Read the docs carefully.

bswinnerton commented:
> I think you need to append the output in listeners.

👍 Good eye. I modified the code to append to two strings: `stdout` and `stderr` in 7e6caca. The local test now outputs:

```
$ INPUT_PLAYBOOK=test/playbook.yml node main.js
[command]/Users/brooks/.dotfiles/bin/ansible-playbook test/playbook.yml
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that
the implicit localhost does not match 'all'

PLAY [Test Action] *************************************************************
skipping: no hosts matched

PLAY RECAP *********************************************************************

::set-output name=stdout::%0APLAY [Test Action] *************************************************************%0Askipping: no hosts matched%0A%0APLAY RECAP *********************************************************************%0A%0A
::set-output name=stderr::[WARNING]: No inventory was parsed, only implicit localhost is available%0A[WARNING]: provided hosts list is empty, only localhost is available. Note that%0Athe implicit localhost does not match 'all'%0A
```

While this works, it makes me think that it might be preferable to not separate stdout from stderr so that you can see the sequential output of both coming back from Ansible. I took a crack at that in d8a8e91. The local test run now outputs:

```
$ INPUT_PLAYBOOK=test/playbook.yml node main.js
[command]/Users/brooks/.dotfiles/bin/ansible-playbook test/playbook.yml
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that
the implicit localhost does not match 'all'

PLAY [Test Action] *************************************************************
skipping: no hosts matched

PLAY RECAP *********************************************************************

::set-output name=output::[WARNING]: No inventory was parsed, only implicit localhost is available%0A[WARNING]: provided hosts list is empty, only localhost is available. Note that%0Athe implicit localhost does not match 'all'%0A%0APLAY [Test Action] *************************************************************%0Askipping: no hosts matched%0A%0APLAY RECAP *********************************************************************%0A%0A
```

bswinnerton commented:
Gave this a shot using my branch and looks like the shifting of `args` may have gone awry:

```
/opt/pipx_bin/ansible-playbook site.yml --tags bird --inventory hosts.yml --extra-vars ansible_user=foo --check --diff --key-file .ansible_key
ERROR! the playbook: --tags bird --inventory hosts.yml --extra-vars ansible_user=foo --check --diff could not be found
Error: The process '/opt/pipx_bin/ansible-playbook' failed with exit code 1
```

Will take another look in the morning.

dawidd6 commented on main.js:
```diff
@@ -16,7 +16,7 @@ async function main() {
         const sudo    = core.getInput("sudo")
 
-        let cmd = ["ansible-playbook", playbook]
+        let cmd = [playbook]
```
Please leave it as it was. Needs to be like that cause later we prepend sudo if needed.

bswinnerton replied:
👍 resolved in bf8069946f8046c1f97e86d007633a15ad6a2e4a.

dawidd6 commented:
Thanks
